File: 623-add-one-row-to-tree/623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if root==None:return root
        if depth==1:
            newNode = TreeNode(val)
            newNode.left=root
            return newNode
        
        def solve(root,val,depth,curDepth):
            if root==None: return
            if depth-1>curDepth:
                solve(root.left,val,depth,curDepth+1)
                solve(root.right,val,depth,curDepth+1)
                return
            newNode1, newNode2 = TreeNode(val), TreeNode(val)
            left,right=root.left,root.right
            root.left,root.right=newNode1,newNode2
            newNode1.left,newNode2.right=left,right
            return
        
        solve(root,val,depth,1)
        return root

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void solve(TreeNode* root, int val, int depth, int curDepth){
        if(!root)return;
        if(depth-1>curDepth){
            solve(root->left,val,depth,curDepth+1);
            solve(root->right,val,depth,curDepth+1);
            return;
        }
        TreeNode* newNode1 = new TreeNode(val);
        TreeNode* newNode2 = new TreeNode(val);
        TreeNode* left = root->left;
        TreeNode* right = root->right;
        root->left=newNode1;
        root->right=newNode2;
        newNode1->left=left;
        newNode2->right=right;
        return;
    }
    
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if(!root)return root;
        if(depth==1){
            TreeNode* newHead = new TreeNode(val);
            newHead->left=root;
            return newHead;
        }
        solve(root,val,depth,1);
        return root;
    }
};